### DISCUSSION TITLE
How can 2 picos communicate using USB

### DISCUSSION THREAD

**Discussion by destatez** [micropython/micropython#13636] [RP2040 / Pico]
I’ve got 2 Picos that are tied to a USB hub, for power. I’d like one of those to transmit text data to the other using the USB. Can I just set Pico1 to use its uart0 for transmit and Pico2 to use its uart1 for receive? I only need communication in this one direction. This seems too simple to get the two Picos talking over USB. Is there something else I should do?

**Comment by destatez** [comment]
This isn’t an answer, it’s a clarification. I have forgotten how to mod the initial discussion. I’m using MicroPython for the programming language for both.

**Reply by GitHubsSilverBullet** [reply]
No you can't.
USB is a hierarchical tree-like bus system and there's  only one master per tree. That's the PC.

You could plug two pi picos together, independently of a USB switch.
One would be programmed as USB host and one as USB device.
But that's not possible using Micropython  (at the moment).
You'd have to program them (at least the USB host) in ›C/C++‹

**Reply by davefes** [reply]
Would it be possible to use a:
CJMCU-312 FT312D USB Android Host UART USB/Serial RS485/RS232

**Reply by GitHubsSilverBullet** [reply]
Had to ›DuckDuckGo‹ that one.
One would have to (at least) implement the ›Android Open Accessory protocol‹ No idea how complex that'd be.
And it would be a 1:1 connection between two devices. No USB-Hub between them.
But a guarantee that it'll work? I'd put significantly more trust in my GCC/G++ and do it myself.

**Edit:**
On second thought, I'd use a fast bidirectional UART connection and be done with it in five minutes.

**Reply by peterhinch** [reply]
> On second thought, I'd use a fast bidirectional UART connection and be done with it in five minutes.

That is definitely the easy way.

**Reply by destatez** [reply]
@peterhinch I would like to get more information on this topic. Would this mean replacing the default Pi Pico UART connections with this fast bidirectional UART? What would the MicroPython code look like? I am concerned with @GitHubsSilverBullet ‘s comments above about requiring a master.

**Reply by GitHubsSilverBullet** [reply]
Connect pin1 on µC1 to pin2 on µC2.
Connect pin2 on µC1 to pin1 on µC2.
If they're plugged into the same USB-hub, they already have a common ground connection.
```python
# on both µCs
from machine import UART
# make this slow enough that your application still works
# yet sensitivity (to spikes and other disturbances) remains low.
u = UART(0, 9600)  # roughly 1KiB/s

# Send data to the other µC
u.write('ABC')

# Read data from the other µC
data = u.read()
```

**Reply by robert-hh** [reply]
> If they're plugged into the same USB-hub, they already have a common ground connection.

I would always add the GND wire and not trust on "it is somehow connected". For DC that may be fine, for anything beyond mains frequency is is a source of trouble. Forgetting it the GND connection a common mistake.

**Reply by GitHubsSilverBullet** [reply]
Good advice. Better double ground than none.

**Reply by GitHubsSilverBullet** [reply]
There seems to be a, hmmm, slight discrepancy between Micropython and RP2040 datasheet.
```python
#!/usr/bin/python3
# -*- coding: UTF-8 -*-
# vim:fileencoding=UTF-8:ts=4

# BRIDGE pin 1 and 2
import re
from machine import UART
from time import sleep_ms, sleep_us

regex = re.compile('.*baudrate=([0-9]*)')
u = UART(0, 125_000_000)
result = regex.match(str(u))
if result:
    result = int(result.group(1))/1e6
    print(f'UART: {result} Mbaud')

data_out = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'
for _ in range(1000):
    u.write(data_out)
    data_in = u.read().decode()
    if data_in != data_out:
        print('ERROR')
```
Spews out 125.0 Mbaud and works … Shouldn't according to the datasheet.
"Supports a maximum baud rate of UARTCLK / 16 in UART mode (7.8 Mbaud at 125MHz)"

**Reply by robert-hh** [reply]
So i assume you made a local loopback between RX an TX. That is the least critical set-up, because is practically works in synchronous mode. Hooking up an oscilloscope shows an actual baud rate of 7,874 MBaud.

**Reply by GitHubsSilverBullet** [reply]
Correct. But shouldn't it throw an exception when configuring a wrong baudrate?
Or accept it but report the true baudrate?

**Reply by robert-hh** [reply]
Both is possible and pretty easy. Someone has to make it, test it an create a PR. That would be the 343th open PR.

Like that, changing line 370 of machine_uart.c into:
`        self->baudrate = uart_init(self->uart, self->baudrate);`
```
>>> u=UART(0, 125_000_000)
>>> u
UART(0, baudrate=7812500, bits=8, parity=None, stop=1, tx=0, rx=1, txbuf=256, rxbuf=256, timeout=0, timeout_char=1, invert=None)
>>> 
```

**Reply by GitHubsSilverBullet** [reply]
Don't get me wrong, I certainly appreciate the amount of work put into Micropython.
But there's a reason why the number of open PRs rises exponentially with the ever increasing number of supported platforms. Sometimes less is more. ;-)

**Reply by robert-hh** [reply]
If one looks into details, simple things are not simple again. Comparing the achieved baud rates with the ones requested, the error is minor over the range from 119 Baud to about 7.8 MBaud. Given that, e.g. reporting 115_199 baud instead of 115_200 seems confusing. So for minor deviations it's better to show the requested rate. But then the change is a little bit more code than a simple assignment. And the two edge cases may not justify that code change, even if it is small.

**Reply by GitHubsSilverBullet** [reply]
I'd like to submit the idea that a requested (and reported) baudrate of 125Mbaud and a real achieved baudrate of 7.8Mbaud is a tad more than a slight/minor difference. ;-)

**Reply by robert-hh** [reply]
@GitHubsSilverBullet Just a note: I added a commit to an existing open PR. With that change, an exception is raised if a  baud rate cannot be set with an error less than about 2%. In my tests, that happens only for too low or too high baud rates. The test covered >1170 baud rates in the range from 120 baud to 7_812_500 baud.

**Reply by GitHubsSilverBullet** [reply]
@robert-hh Thanks for that. 
As a matter of fact I found something far more serious in the UART part of the ›machine‹ implementation. 
Rather interesting that apparently nobody stumbled over it.
I'm currently traveling, so have no lab equipment. Just a PiPico dangling from the USB cable. But that's plenty enough to demonstrate. Wrote a logger with single cycle resolution (using the PIO).
Need another day or two and will then provide a complete test program (might be lengthy, so it needs some documentation)
Btw. Tested on 1.21.1

**Reply by robert-hh** [reply]
1.21.0 or 1.22.1? You should test with 1.22.1.

**Reply by GitHubsSilverBullet** [reply]
Sorry. Typo!
MPY: soft reboot
MicroPython v1.22.1 on 2024-01-05; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>>
